File: app/src/main/java/me/devsaki/hentoid/parsers/images/PixivParser.java
package me.devsaki.hentoid.parsers.images;

import static me.devsaki.hentoid.util.ContentHelper.KEY_DL_PARAMS_NB_CHAPTERS;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.annimon.stream.Stream;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import me.devsaki.hentoid.database.domains.Attribute;
import me.devsaki.hentoid.database.domains.Chapter;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.enums.Site;
import me.devsaki.hentoid.enums.StatusContent;
import me.devsaki.hentoid.json.sources.PixivIllustMetadata;
import me.devsaki.hentoid.json.sources.PixivIllustPagesMetadata;
import me.devsaki.hentoid.json.sources.PixivSeriesIllustMetadata;
import me.devsaki.hentoid.json.sources.PixivUserIllustMetadata;
import me.devsaki.hentoid.parsers.ParseHelper;
import me.devsaki.hentoid.retrofit.sources.PixivServer;
import me.devsaki.hentoid.util.ContentHelper;
import me.devsaki.hentoid.util.StringHelper;
import me.devsaki.hentoid.util.download.DownloadRateLimiter;
import me.devsaki.hentoid.util.exception.EmptyResultException;
import me.devsaki.hentoid.util.exception.PreparationInterruptedException;
import me.devsaki.hentoid.util.network.HttpHelper;
import retrofit2.Response;
import timber.log.Timber;

public class PixivParser extends BaseImageListParser {

    private static final int MAX_QUERY_WINDOW = 30;

    @Override
    protected List<ImageFile> parseImageListImpl(@NonNull Content onlineContent, @Nullable Content storedContent) throws Exception {
        EventBus.getDefault().register(this);
        processedUrl = onlineContent.getGalleryUrl();

        List<ImageFile> result;
        try {
            result = getPages(onlineContent, storedContent);
        } finally {
            EventBus.getDefault().unregister(this);
        }

        return result;
    }

    private List<ImageFile> getPages(@NonNull Content onlineContent, @Nullable Content storedContent) throws Exception {

        try {
            boolean useMobileAgent = Site.PIXIV.useMobileAgent();
            boolean useHentoidAgent = Site.PIXIV.useHentoidAgent();
            boolean useWebviewAgent = Site.PIXIV.useWebviewAgent();

            String cookieStr = HttpHelper.getCookies(
                    onlineContent.getGalleryUrl(), null,
                    useMobileAgent, useHentoidAgent, useWebviewAgent);

            // API calls seem to be protected against request spam; 1 is arbitrary
            DownloadRateLimiter.INSTANCE.setRateLimit(1);

            if (onlineContent.getUrl().contains("/series/"))
                return parseSeries(onlineContent, storedContent, cookieStr);
            else if (onlineContent.getUrl().contains("/artworks/"))
                return parseIllust(onlineContent, cookieStr);
            else if (onlineContent.getUrl().contains("users/"))
                return parseUser(onlineContent, storedContent, cookieStr);
        } catch (Exception e) {
            Timber.d(e);
            throw new EmptyResultException(StringHelper.protect(e.getMessage()));
        }
        return Collections.emptyList();
    }

    private List<ImageFile> parseIllust(@NonNull Content content, @NonNull String cookieStr) throws Exception {
        DownloadRateLimiter.INSTANCE.take();
        PixivIllustPagesMetadata galleryMetadata = PixivServer.api.getIllustPages(content.getUniqueSiteId(), cookieStr).execute().body();
        if (null == galleryMetadata || galleryMetadata.isError()) {
            String message = "";
            if (galleryMetadata != null) message = galleryMetadata.getMessage();
            throw new EmptyResultException(message);
        }
        return ParseHelper.urlsToImageFiles(galleryMetadata.getPageUrls(), content.getCoverImageUrl(), StatusContent.SAVED);
    }

    private List<ImageFile> parseSeries(@NonNull Content onlineContent, @Nullable Content storedContent, @NonNull String cookieStr) throws Exception {
        String[] seriesIdParts = onlineContent.getUniqueSiteId().split("/");
        String seriesId = seriesIdParts[seriesIdParts.length - 1];
        if (seriesId.contains("?")) {
            seriesId = seriesId.substring(0, seriesId.indexOf("?"));
        }

        // Retrieve the number of Illusts
        String nbChaptersStr = ContentHelper.parseDownloadParams(onlineContent.getDownloadParams()).get(KEY_DL_PARAMS_NB_CHAPTERS);
        if (null == nbChaptersStr || !StringHelper.isNumeric(nbChaptersStr))
            throw new IllegalArgumentException("Chapter count not saved");
        int nbChapters = Integer.parseInt(nbChaptersStr);

        // List all Illust IDs (API is paged, hence the loop)
        List<Chapter> chapters = new ArrayList<>();
        while (chapters.size() < nbChapters) {
            DownloadRateLimiter.INSTANCE.take();
            if (processHalted.get()) break;
            int chaptersToRead = Math.min(nbChapters - chapters.size(), MAX_QUERY_WINDOW);
            DownloadRateLimiter.INSTANCE.take();
            PixivSeriesIllustMetadata seriesContentMetadata = PixivServer.api.getSeriesIllusts(seriesId, chaptersToRead, chapters.size(), cookieStr).execute().body();
            if (null == seriesContentMetadata || seriesContentMetadata.isError()) {
                String message = "Unreachable series illust";
                if (seriesContentMetadata != null)
                    message = seriesContentMetadata.getMessage();
                throw new IllegalArgumentException(message);
            }
            chapters.addAll(seriesContentMetadata.getChapters(onlineContent.getId()));
        }
        // Put back chapters in reading order
        Collections.reverse(chapters);


        // If the stored content has chapters already, save them for comparison
        List<Chapter> storedChapters = null;
        if (storedContent != null) {
            storedChapters = storedContent.getChapters();
            if (storedChapters != null)
                storedChapters = Stream.of(storedChapters).toList(); // Work on a copy
        }
        if (null == storedChapters) storedChapters = Collections.emptyList();

        // Use chapter folder as a differentiator (as the whole URL may evolve)
        List<Chapter> extraChapters = ParseHelper.getExtraChaptersbyUrl(storedChapters, chapters);

        progressStart(onlineContent, storedContent, extraChapters.size());

        // Start numbering extra images right after the last position of stored and chaptered images
        int imgOffset = ParseHelper.getMaxImageOrder(storedChapters) + 1;

        // Retrieve all Illust detailed info
        List<ImageFile> result = new ArrayList<>();
        result.add(ImageFile.newCover(onlineContent.getCoverImageUrl(), StatusContent.SAVED));
        Set<Attribute> attrs = new HashSet<>();
        for (Chapter ch : extraChapters) {
            DownloadRateLimiter.INSTANCE.take();
            PixivIllustMetadata illustMetadata = PixivServer.api.getIllustMetadata(ch.getUniqueId(), cookieStr).execute().body();
            if (null == illustMetadata || illustMetadata.isError()) {
                String message = "Unreachable illust";
                if (illustMetadata != null)
                    message = illustMetadata.getMessage();
                throw new IllegalArgumentException(message);
            }

            List<Attribute> chapterAttrs = illustMetadata.getAttributes();
            attrs.addAll(chapterAttrs);

            List<ImageFile> chapterImages = illustMetadata.getImageFiles();
            for (ImageFile img : chapterImages)
                img.setOrder(imgOffset++).computeName(4).setChapter(ch);

            result.addAll(chapterImages);

            if (processHalted.get()) break;
            progressPlus();
        }
        // If the process has been halted manually, the result is incomplete and should not be returned as is
        if (processHalted.get()) throw new PreparationInterruptedException();

        onlineContent.putAttributes(attrs);
        onlineContent.setUpdatedProperties(true);

        progressComplete();
        return result;
    }

    private List<ImageFile> parseUser(@NonNull Content onlineContent, @Nullable Content storedContent, @NonNull String cookieStr) throws Exception {
        String[] userIdParts = onlineContent.getUniqueSiteId().split("/");
        String userId = userIdParts[userIdParts.length - 1];
        if (userId.contains("?")) {
            userId = userId.substring(0, userId.indexOf("?"));
        }

        // Retrieve the list of Illusts IDs (=chapters)
        DownloadRateLimiter.INSTANCE.take();
        PixivUserIllustMetadata userIllustsMetadata = PixivServer.api.getUserIllusts(userId, cookieStr).execute().body();
        if (null == userIllustsMetadata || userIllustsMetadata.isError()) {
            String message = "Unreachable user illusts";
            if (userIllustsMetadata != null)
                message = userIllustsMetadata.getMessage();
            throw new IllegalArgumentException(message);
        }

        // Detect extra chapters
        List<String> illustIds = userIllustsMetadata.getIllustIds();
        List<Chapter> storedChapters = null;
        if (storedContent != null) {
            storedChapters = storedContent.getChapters();
            if (storedChapters != null) storedChapters = Stream.of(storedChapters).toList();
        }
        if (null == storedChapters) storedChapters = Collections.emptyList();
        else illustIds = ParseHelper.getExtraChaptersbyId(storedChapters, illustIds);

        // Work on detected extra chapters
        progressStart(onlineContent, storedContent, illustIds.size());

        // Start numbering extra images & chapters right after the last position of stored and chaptered images & chapters
        int imgOffset = ParseHelper.getMaxImageOrder(storedChapters) + 1;
        int chpOffset = ParseHelper.getMaxChapterOrder(storedChapters) + 1;

        // Cycle through all Illusts
        List<ImageFile> result = new ArrayList<>();
        result.add(ImageFile.newCover(onlineContent.getCoverImageUrl(), StatusContent.SAVED));
        Set<Attribute> attrs = new HashSet<>();
        for (String illustId : illustIds) {
            DownloadRateLimiter.INSTANCE.take();
            Response<PixivIllustMetadata> resp = PixivServer.api.getIllustMetadata(illustId, cookieStr).execute();
            if (resp.code() >= 400)
                throw new IllegalArgumentException(String.format("Unreachable illust : code=%s (%s)", resp.code(), resp.message()));

            PixivIllustMetadata illustMetadata = resp.body();
            if (null == illustMetadata || illustMetadata.isError()) {
                String message;
                if (illustMetadata != null)
                    message = illustMetadata.getMessage();
                else
                    message = "no metadata";
                throw new IllegalArgumentException(String.format("Unreachable illust : %s", message));
            }

            List<Attribute> chapterAttrs = illustMetadata.getAttributes();
            attrs.addAll(chapterAttrs);

            Chapter chp = new Chapter(chpOffset++, illustMetadata.getUrl(), illustMetadata.getTitle()).setUniqueId(illustMetadata.getId()).setContentId(onlineContent.getId());
            List<ImageFile> chapterImages = illustMetadata.getImageFiles();
            for (ImageFile img : chapterImages)
                img.setOrder(imgOffset++).computeName(4).setChapter(chp);

            result.addAll(chapterImages);

            if (processHalted.get()) break;
            progressPlus();
        }
        // If the process has been halted manually, the result is incomplete and should not be returned as is
        if (processHalted.get()) throw new PreparationInterruptedException();

        onlineContent.putAttributes(attrs);
        onlineContent.putAttributes(attrs);
        onlineContent.setUpdatedProperties(true);

        progressComplete();
        return result;
    }

    @Override
    protected List<String> parseImages(@NonNull Content content) {
        /// We won't use that as parseImageList is overriden directly
        return null;
    }
}

File: app/src/main/java/me/devsaki/hentoid/parsers/images/PixivParser.kt
package me.devsaki.hentoid.parsers.images

import me.devsaki.hentoid.database.domains.Attribute
import me.devsaki.hentoid.database.domains.Chapter
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.enums.StatusContent
import me.devsaki.hentoid.parsers.getExtraChaptersbyId
import me.devsaki.hentoid.parsers.getExtraChaptersbyUrl
import me.devsaki.hentoid.parsers.getMaxChapterOrder
import me.devsaki.hentoid.parsers.getMaxImageOrder
import me.devsaki.hentoid.parsers.getUserAgent
import me.devsaki.hentoid.parsers.urlsToImageFiles
import me.devsaki.hentoid.retrofit.sources.PixivServer
import me.devsaki.hentoid.util.KEY_DL_PARAMS_NB_CHAPTERS
import me.devsaki.hentoid.util.Settings
import me.devsaki.hentoid.util.download.DownloadRateLimiter.setRateLimit
import me.devsaki.hentoid.util.download.DownloadRateLimiter.take
import me.devsaki.hentoid.util.exception.EmptyResultException
import me.devsaki.hentoid.util.exception.PreparationInterruptedException
import me.devsaki.hentoid.util.isNumeric
import me.devsaki.hentoid.util.network.getCookies
import me.devsaki.hentoid.util.network.waitBlocking429
import me.devsaki.hentoid.util.parseDownloadParams
import org.greenrobot.eventbus.EventBus
import timber.log.Timber

private const val MAX_QUERY_WINDOW = 30

class PixivParser : BaseImageListParser() {

    override fun parseImages(content: Content): List<String> {
        // We won't use that as parseImageList is overriden directly
        return emptyList()
    }

    @Throws(Exception::class)
    override fun parseImageListImpl(
        onlineContent: Content,
        storedContent: Content?
    ): List<ImageFile> {
        EventBus.getDefault().register(this)
        processedUrl = onlineContent.galleryUrl
        val result: List<ImageFile> = try {
            getPages(onlineContent, storedContent)
        } finally {
            EventBus.getDefault().unregister(this)
        }
        return result
    }

    @Throws(Exception::class)
    private fun getPages(onlineContent: Content, storedContent: Content?): List<ImageFile> {
        try {
            val useMobileAgent = Site.PIXIV.useMobileAgent
            val useHentoidAgent = Site.PIXIV.useHentoidAgent
            val useWebviewAgent = Site.PIXIV.useWebviewAgent
            val cookieStr = getCookies(
                onlineContent.galleryUrl, null,
                useMobileAgent, useHentoidAgent, useWebviewAgent
            )
            val userAgent = getUserAgent(Site.PIXIV)
            val acceptAll = "*/*"

            // API calls seem to be protected against request spam; 2 is arbitrary
            setRateLimit(2)
            if (onlineContent.url.contains("/series/")) return parseSeries(
                onlineContent,
                storedContent,
                cookieStr,
                acceptAll,
                userAgent
            ) else if (onlineContent.url.contains("/artworks/")) return parseIllust(
                onlineContent,
                cookieStr,
                acceptAll,
                userAgent
            ) else if (onlineContent.url.contains("users/")) return parseUser(
                onlineContent,
                storedContent,
                cookieStr,
                acceptAll,
                userAgent
            )
        } catch (e: Exception) {
            Timber.d(e)
            throw EmptyResultException(e.message ?: "")
        }
        take() // One last delay before download phase
        return emptyList()
    }

    @Throws(Exception::class)
    private fun parseIllust(
        content: Content,
        cookieStr: String,
        acceptAll: String,
        userAgent: String
    ): List<ImageFile> {
        take()
        val galleryMetadata =
            PixivServer.api.getIllustPages(content.uniqueSiteId, cookieStr, acceptAll, userAgent)
                .execute().body()
        if (null == galleryMetadata || galleryMetadata.error == true) {
            var message = ""
            if (galleryMetadata != null) message = galleryMetadata.message ?: ""
            throw EmptyResultException(message)
        }
        return urlsToImageFiles(
            galleryMetadata.getPageUrls(),
            content.coverImageUrl,
            StatusContent.SAVED
        )
    }

    @Throws(Exception::class)
    private fun parseSeries(
        onlineContent: Content,
        storedContent: Content?,
        cookieStr: String,
        acceptAll: String,
        userAgent: String
    ): List<ImageFile> {
        val seriesIdParts =
            onlineContent.uniqueSiteId.split("/")
        var seriesId = seriesIdParts[seriesIdParts.size - 1]
        if (seriesId.contains("?")) {
            seriesId = seriesId.substring(0, seriesId.indexOf("?"))
        }

        // Retrieve the number of Illusts
        val nbChaptersStr =
            parseDownloadParams(onlineContent.downloadParams)[KEY_DL_PARAMS_NB_CHAPTERS]
        require(nbChaptersStr != null) { "Chapter count not saved" }
        require(isNumeric(nbChaptersStr)) { "Chapter count not saved" }
        val nbChapters = nbChaptersStr.toInt()

        // List all Illust IDs (API is paged, hence the loop)
        val chapters: MutableList<Chapter> = ArrayList()
        while (chapters.size < nbChapters) {
            if (processHalted.get()) break
            val chaptersToRead = (nbChapters - chapters.size).coerceAtMost(MAX_QUERY_WINDOW)
            take()
            val seriesContentMetadata = PixivServer.api.getSeriesIllusts(
                seriesId,
                chaptersToRead,
                chapters.size,
                cookieStr,
                acceptAll,
                userAgent
            ).execute().body()
            if (null == seriesContentMetadata || true == seriesContentMetadata.error) {
                var message: String? = "Unreachable series illust"
                if (seriesContentMetadata != null) message = seriesContentMetadata.message
                throw IllegalArgumentException(message)
            }
            chapters.addAll(seriesContentMetadata.getChapters(onlineContent.id))
        }
        // Put back chapters in reading order
        chapters.reverse()


        // If the stored content has chapters already, save them for comparison
        var storedChapters: List<Chapter>? = null
        if (storedContent != null) {
            storedChapters = storedContent.chapters
            storedChapters = storedChapters.toMutableList() // Work on a copy
        }
        if (null == storedChapters) storedChapters = emptyList()

        // Use chapter folder as a differentiator (as the whole URL may evolve)
        val extraChapters = getExtraChaptersbyUrl(storedChapters, chapters)
        progressStart(onlineContent, storedContent)

        // Start numbering extra images right after the last position of stored and chaptered images
        var imgOffset = getMaxImageOrder(storedChapters) + 1

        // Retrieve all Illust detailed info
        val result: MutableList<ImageFile> = ArrayList()
        result.add(ImageFile.newCover(onlineContent.coverImageUrl, StatusContent.SAVED))
        val attrs: MutableSet<Attribute> = HashSet()
        extraChapters.forEachIndexed { index, ch ->
            if (processHalted.get()) return@forEachIndexed
            take()
            val illustMetadata =
                PixivServer.api.getIllustMetadata(ch.uniqueId, cookieStr, acceptAll, userAgent)
                    .execute().body()
            if (null == illustMetadata || illustMetadata.error) {
                var message: String? = "Unreachable illust"
                if (illustMetadata != null) message = illustMetadata.message
                throw IllegalArgumentException(message)
            }
            val chapterAttrs = illustMetadata.getAttributes()
            attrs.addAll(chapterAttrs)
            val chapterImages = illustMetadata.getImageFiles()
            for (img in chapterImages) {
                img.order = imgOffset++
                img.setChapter(ch)
                img.computeName(4)
            }
            result.addAll(chapterImages)
            progressPlus((index + 1f) / extraChapters.size)
        }
        // If the process has been halted manually, the result is incomplete and should not be returned as is
        if (processHalted.get()) throw PreparationInterruptedException()
        onlineContent.putAttributes(attrs)
        progressComplete()
        return result
    }

    @Throws(Exception::class)
    private fun parseUser(
        onlineContent: Content,
        storedContent: Content?,
        cookieStr: String,
        acceptAll: String,
        userAgent: String
    ): List<ImageFile> {
        val userIdParts =
            onlineContent.uniqueSiteId.split("/")
        var userId = userIdParts[userIdParts.size - 1]
        if (userId.contains("?")) {
            userId = userId.substring(0, userId.indexOf("?"))
        }

        // Retrieve the list of Illusts IDs (=chapters)
        var waited = 0
        take()
        var userIllustResp =
            PixivServer.api.getUserIllusts(userId, cookieStr, acceptAll, userAgent).execute()
        if (waitBlocking429(
                userIllustResp,
                Settings.http429DefaultDelaySecs * 1000
            )
        ) {
            waited++
            userIllustResp =
                PixivServer.api.getUserIllusts(userId, cookieStr, acceptAll, userAgent).execute()
        }
        require(userIllustResp.code() < 400) {
            String.format(
                "Unreachable user illusts : code=%s (%s) [%d]",
                userIllustResp.code(),
                userIllustResp.message(),
                waited
            )
        }
        val userIllustsMetadata = userIllustResp.body()
        if (null == userIllustsMetadata || userIllustsMetadata.isError()) {
            var message: String? = "Unreachable user illusts"
            if (userIllustsMetadata != null) message = userIllustsMetadata.getMessage()
            throw IllegalArgumentException(message)
        }

        // Detect extra chapters
        var illustIds = userIllustsMetadata.getIllustIds()
        var storedChapters: List<Chapter>? = null
        if (storedContent != null) {
            storedChapters = storedContent.chapters
            storedChapters = storedChapters.toMutableList()
        }
        if (null == storedChapters) storedChapters = emptyList()
        else illustIds = getExtraChaptersbyId(storedChapters, illustIds)

        // Work on detected extra chapters
        progressStart(onlineContent, storedContent)

        // Start numbering extra images & chapters right after the last position of stored and chaptered images & chapters
        var imgOffset = getMaxImageOrder(storedChapters) + 1
        var chpOffset = getMaxChapterOrder(storedChapters) + 1

        // Cycle through all Illusts
        val result: MutableList<ImageFile> = ArrayList()
        result.add(ImageFile.newCover(onlineContent.coverImageUrl, StatusContent.SAVED))
        val attrs: MutableSet<Attribute> = HashSet()
        illustIds.forEachIndexed { index, illustId ->
            waited = 0
            take()
            if (processHalted.get()) return@forEachIndexed
            var illustResp =
                PixivServer.api.getIllustMetadata(illustId, cookieStr, acceptAll, userAgent)
                    .execute()
            while (waitBlocking429(
                    illustResp,
                    Settings.http429DefaultDelaySecs * 1000
                ) && waited < 2
            ) {
                waited++
                illustResp =
                    PixivServer.api.getIllustMetadata(illustId, cookieStr, acceptAll, userAgent)
                        .execute()
            }
            require(illustResp.code() < 400) {
                String.format(
                    "Unreachable illust : code=%s (%s) [%d - %d]",
                    illustResp.code(),
                    illustResp.message(),
                    index,
                    waited
                )
            }
            val illustMetadata = illustResp.body()
            if (null == illustMetadata || illustMetadata.error) {
                val message: String? =
                    if (illustMetadata != null) illustMetadata.message else "no metadata"
                throw IllegalArgumentException(String.format("Unreachable illust : %s", message))
            }
            val chapterAttrs = illustMetadata.getAttributes()
            attrs.addAll(chapterAttrs)
            val chp = Chapter(
                order = chpOffset++,
                url = illustMetadata.getUrl(),
                name = illustMetadata.getTitle(),
                uniqueId = illustMetadata.getId()
            )
            chp.setContentId(onlineContent.id)
            val chapterImages = illustMetadata.getImageFiles()
            for (img in chapterImages) {
                img.order = imgOffset++
                img.setChapter(chp)
                img.computeName(4)
            }
            result.addAll(chapterImages)
            progressPlus((index + 1f) / illustIds.size)
        }
        // If the process has been halted manually, the result is incomplete and should not be returned as is
        if (processHalted.get()) throw PreparationInterruptedException()
        onlineContent.putAttributes(attrs)
        onlineContent.putAttributes(attrs)
        progressComplete()
        return result
    }
}